### DISCUSSION TITLE
Add friction to the terrain surface

### DISCUSSION THREAD

**Discussion by Alibutter** [bulletphysics/bullet3#3961] [Q&A]
I am trying to build my own gym environment with PyBullet. I used the following code to load the 3d terrain stored in the TXT file.  Normally, the car would be stationary on the slope because of friction. But when I tested this gym environment, the car always skidded on the slope, except when it hit a gully.
```
def reset(self):
    p.configureDebugVisualizer(p.COV_ENABLE_GUI, 0)
    p.resetSimulation(physicsClientId=self._physics_client_id)
    p.setGravity(0, 0, -1)
    
    # height data path
    file_name = 'N21E099'
    terrain_csv_path = DataConfig.TERRAIN_PATH + '{}_{}x{}.txt'.format(file_name, self.map_size, self.map_size)

    # write height data to txt file
    if not os.path.exists(terrain_csv_path):
        self.dg.init_dem(DataConfig.DEM_PATH + file_name + '.hgt')
        height = np.float32(self.dg.get_original() / 50)
        np.savetxt(terrain_csv_path, height, delimiter=',')

    # read the txt contains height data of the 3D terrain
    terrain_shape = p.createCollisionShape(
        shapeType=p.GEOM_HEIGHTFIELD,
        meshScale=[.1, .1, .1],
        fileName=terrain_csv_path,
        heightfieldTextureScaling=512
    )

    terrain = p.createMultiBody(0, terrain_shape)
    p.resetBasePositionAndOrientation(terrain, [0, 0, 0], [0, 0, 0, 1])
    p.changeVisualShape(terrain, -1, rgbaColor=[1, 1, 1, 1])

    # load car model
    self.car = p.loadURDF(
        fileName=DataConfig.URDF_PATH + 'mycar.urdf',
        basePosition=[0., 0., 0],
        physicsClientId=self._physics_client_id
    )
    return self.__get_observation()
```
![surface](https://user-images.githubusercontent.com/49897267/133450941-df02a9f4-2f7a-4f03-9651-4110eff20385.png)

https://user-images.githubusercontent.com/49897267/133449911-64f2ed5d-c86f-46a5-8e30-cae8077e368b.mp4

I spent a lot of time, but no one encountered the problem of the car slipping on the sloped terrain. I tried to reduce the gravity to 1, and reduced the mass of the car. The problem still exists.
The most important thing is that my terrain is loaded from a TXT file, and pybullet does not seem to provide an interface for adding friction to the terrain surface under this loading method.

So how can I prevent the  car from slipping on the slope? Does pybullet provide other methods to add the friction to the surface of CollisionShape(by using p.createCollisionShape()) or MultiBody(by using p.createMultiBody()) ?The solution of this problem is of great help to me，I would be very grateful if anyone can solve this problem.

**Comment by Hopson97** [comment]
Hi I know this is very late but for anyone who comes across this question this might be very useful.

When creating a RigidBody in Bullet, the friction can be set on the `btRigidBody::btRigidBodyConstructionInfo`

In C++ this is done by doing something like this:

```cpp
    // ...
    // ... Create your collision shape etc here ...
    // ...


    // Create the rigid body for the ground
    btDefaultMotionState* ground_motion_state = new btDefaultMotionState(ground_transform_bt);

    // Create the rigid body construction info
    btRigidBody::btRigidBodyConstructionInfo ground_rb_info(
        mass, ground_motion_state, ground_shape, ground_shape_local_inertia);
   
    // This is what sets the friction for the ground, I find 2.0 is quite nice for "grass"
    // Bullet sets this to 0.5 by default
    ground_rb_info.m_friction = 2.0f; 
    
    // Actually create the rigid body etc
    btRigidBody* ground_rb = new btRigidBody(ground_rb_info);
    dynamics_world.addRigidBody(ground_rb);
```

I'm not sure how this works in PyBullet, but I assume it is very similar
